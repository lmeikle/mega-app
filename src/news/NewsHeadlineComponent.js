import React from 'react';
import { string } from 'prop-types';
import './News.css';

/**
 * Stateless component which renders a single news headline.
 */
const NewsHeadlineComponent = ({ title, date, source, url }) => {
  return (
    <div className="news-headline">
      <a href={url} target="_blank" rel="noopener noreferrer">
        <div className="news-headline-title">{title}</div>
      </a>
      <div className="news-headline-date-and-source-container">
        <div className="news-headline-date">{date}</div>
        <div className="news-headline-source">{source}</div>
      </div>
    </div>
  );
};

NewsHeadlineComponent.propTypes = {
  title: string.isRequired,
  date: string.isRequired,
  source: string.isRequired,
  url: string.isRequired
};

NewsHeadlineComponent.defaultProps = {};

export default NewsHeadlineComponent;
