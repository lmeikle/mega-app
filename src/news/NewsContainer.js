import React, { Component } from 'react';
import { fetchTopHeadlines } from './NewsAPI';
import NewsHeadlineComponent from './NewsHeadlineComponent';
import LoadingComponent from '../shared/loading/LoadingComponent';

/**
 * Renders a list of top news headlines
 */
class NewsContainer extends Component {
  constructor(props) {
    super(props);

    this.state = {
      totalResults: 0,
      headlines: [],
      error: null
    };
  }

  static get PAGE_SIZE() {
    return 5;
  }

  componentDidMount() {
    this.showMore();
  }

  showMore = async () => {
    let { totalResults, headlines } = this.state;
    let page = ((headlines.length / totalResults) * totalResults) / NewsContainer.PAGE_SIZE;

    try {
      const result = await fetchTopHeadlines(page + 1, NewsContainer.PAGE_SIZE);
      this.setState({
        headlines: [...this.state.headlines, ...result.headlines],
        totalResults: result.totalResults
      });
    } catch (error) {
      this.setState({ error });
    }
  };

  render() {
    const { headlines, totalResults, error } = this.state;

    if (error) {
      return <div className="errorMessage">Failed to get top headlines due to: {error.toString()}</div>;
    }

    if (totalResults === 0) {
      return <LoadingComponent />;
    }

    return (
      <div className="news-container">
        <div className="news-title">News</div>
        <div className="news-headlines-container">{headlines.map(headline => <NewsHeadlineComponent key={headline.title} {...headline} />)}</div>
        {headlines.length < totalResults && (
          <div className="news-show-more-button" onClick={this.showMore}>
            Show More
          </div>
        )}
        <div className="news-powered-by">
          <a href="https://newsapi.org" target="_blank" rel="noopener noreferrer">
            Powered by News API
          </a>
        </div>
      </div>
    );
  }
}

export default NewsContainer;
