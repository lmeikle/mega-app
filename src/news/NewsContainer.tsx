import * as React from 'react';
import { LoadingComponent } from '@lmeikle/my-mono-repo-to-single-package';
import { fetchTopHeadlines } from './NewsAPI';
import NewsHeadlineComponent from './NewsHeadlineComponent';

type IProps = {};

type IState = {
  totalResults: number;
  headlines: Array<IHeadLineProps>;
  error?: string;
};

export type IHeadLineProps = {
  title: string;
  date: string;
  source: string;
  url: string;
};

/**
 * Renders a list of top news headlines
 */
class NewsContainer extends React.Component<IProps, IState> {
  state: IState = {
    totalResults: 0,
    headlines: []
  };

  static get PAGE_SIZE() {
    return 5;
  }

  componentDidMount() {
    this.showMore();
  }

  showMore = async () => {
    const { totalResults, headlines } = this.state;
    let page = ((headlines.length / totalResults) * totalResults) / NewsContainer.PAGE_SIZE;

    try {
      const result = await fetchTopHeadlines(page + 1, NewsContainer.PAGE_SIZE);
      this.setState({
        headlines: [...this.state.headlines, ...result.headlines],
        totalResults: result.totalResults
      });
    } catch (error) {
      this.setState({ error });
    }
  };

  render() {
    const { headlines, totalResults, error } = this.state;

    if (error) {
      return <div className="errorMessage">Failed to get top headlines due to: {error.toString()}</div>;
    }

    if (totalResults === 0) {
      return <LoadingComponent />;
    }

    return (
      <div className="news-container">
        <div className="news-title">News</div>
        <div className="news-headlines-container">
          {headlines.map(headline => (
            <NewsHeadlineComponent key={headline.title} {...headline} />
          ))}
        </div>
        {headlines.length < totalResults && (
          <div className="news-show-more-button" onClick={this.showMore}>
            Show More
          </div>
        )}
        <div className="news-powered-by">
          <a href="https://newsapi.org" target="_blank" rel="noopener noreferrer">
            Powered by News API
          </a>
        </div>
      </div>
    );
  }
}

export default NewsContainer;
