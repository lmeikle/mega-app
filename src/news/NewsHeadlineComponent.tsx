import React from 'react';
import { IHeadLineProps } from './NewsContainer';
import './News.css';

/**
 * Stateless component which renders a single news headline.
 */
const NewsHeadlineComponent = ({ title, date, source, url }: IHeadLineProps) => {
  return (
    <div className="news-headline">
      <a href={url} target="_blank" rel="noopener noreferrer">
        <div data-testid="NewsHeadlineComponent_title" className="news-headline-title">
          {title}
        </div>
      </a>
      <div className="news-headline-date-and-source-container">
        <div className="news-headline-date">{date}</div>
        <div className="news-headline-source">{source}</div>
      </div>
    </div>
  );
};

export default NewsHeadlineComponent;
