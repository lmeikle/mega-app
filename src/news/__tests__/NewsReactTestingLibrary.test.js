import React from 'react';
import { render, fireEvent, cleanup, waitForElement, wait } from 'react-testing-library';
import 'jest-dom/extend-expect';
import { shallow } from 'enzyme';
import NewsContainer from '../NewsContainer';
import NewsHeadlineComponent from '../NewsHeadlineComponent';
import mockResponse from './mockResponse.json';
import mockResponse2 from './mockResponse2.json';

// automatically unmount and cleanup DOM after the test is finished.
afterEach(cleanup);

describe('News', () => {
  beforeEach(() => {
    fetch.resetMocks();
  });

  /**
   * Things I want to test:
   *
   * -shows the LoadingComponent whilst waiting for data
   * -when mounted makes a fetch request and displays the results
   * -displays the show more button when there is more data
   * -does not display the show more button when there is not more data
   * -renders an error message correctly
   *
   */
  test('shows the LoadingComponent whilst waiting for data', () => {
    // can't use reacct-testing-library for this one
    let newsContainer = shallow(<NewsContainer />);
    expect(newsContainer.find('LoadingComponent').length).toBe(1);
    expect(newsContainer.find('NewsHeadlineComponent').length).toBe(0);
    expect(newsContainer.find('.news-show-more-button').length).toBe(0);
  });

  test('when mounted makes a fetch request and displays the results', async () => {
    fetch.mockResponseOnce(JSON.stringify(mockResponse));

    const { getByText, getByTestId, queryByText, queryAllByTestId, container } = render(<NewsContainer />);

    // Let's wait until our mocked `get` request promise resolves and the component calls setState and re-renders.
    await wait(() => expect(queryByText(mockResponse.articles[0].title)).toBeInTheDocument());

    // check the correct number of NewsHeadlineComponent's have rendered
    expect(queryAllByTestId('NewsHeadlineComponent_title')).toHaveLength(mockResponse.articles.length);

    // check fetch calls are correct, these 2 test the same thing
    expect(fetch.mock.calls.length).toEqual(1);
    expect(fetch).toHaveBeenCalledTimes(1);

    // snapshots work great with regular DOM nodes!
    expect(container.firstChild).toMatchSnapshot();
  });

  test('displays the show more button when there is more data', async () => {
    fetch.mockResponseOnce(JSON.stringify(mockResponse));

    const { getByText, queryByText } = render(<NewsContainer />);

    // Let's wait until our mocked `get` request promise resolves and the component calls setState and re-renders.
    await wait(() => expect(queryByText(mockResponse.articles[0].title)).toBeInTheDocument());

    // check the show more button is visible
    expect(getByText('Show More')).toBeInTheDocument();
  });

  test('does not display the show more button when there is not more data', async () => {
    fetch.mockResponseOnce(JSON.stringify(mockResponse));

    const { getByText, queryByText } = render(<NewsContainer />);

    // Let's wait until our mocked `get` request promise resolves and the component calls setState and re-renders.
    await wait(() => expect(queryByText(mockResponse.articles[0].title)).toBeInTheDocument());

    fetch.mockResponseOnce(JSON.stringify(mockResponse2));
    fireEvent.click(getByText('Show More'));

    // Let's wait until our mocked `get` request promise resolves and the component calls setState and re-renders.
    await wait(() => expect(queryByText(mockResponse2.articles[0].title)).toBeInTheDocument());

    // check the show more button is not visible
    expect(queryByText('Show More')).not.toBeInTheDocument();
  });

  test('renders an error message correctly', async () => {
    const errorMessage = 'foo';
    fetch.mockReject(errorMessage);

    const { getByText, queryByText } = render(<NewsContainer />);

    // Let's wait until our mocked `get` request promise resolves and the component calls setState and re-renders.
    await wait(() => expect(getByText(/Something went wrong .../)).toBeInTheDocument());

    // check the show more button is not visible
    expect(queryByText('Show More')).not.toBeInTheDocument();
  });
});

describe('NewsHeadlineComponent', () => {
  test('renders correctly', () => {
    let props = {
      title: 'Putin ready to meet Trump over arms race concerns',
      date: '10/06/2018',
      source: 'CNN',
      url: 'https://www.cnn.com/2018/06/10/politics/putin-trump-us-russia-arms-intl/index.html'
    };

    const { getByText, getByTestId, container } = render(<NewsHeadlineComponent {...props} />);

    // these 2 test the same thing
    expect(getByTestId('NewsHeadlineComponent_title')).toHaveTextContent(props.title);
    expect(getByText(props.title)).toBeInTheDocument();

    // snapshots work great with regular DOM nodes!
    expect(container.firstChild).toMatchSnapshot();
  });
});
