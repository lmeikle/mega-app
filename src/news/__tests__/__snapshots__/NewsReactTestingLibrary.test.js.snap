// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`News when mounted makes a fetch request and displays the results 1`] = `
<div
  class="news-container"
>
  <div
    class="news-title"
  >
    News
  </div>
  <div
    class="news-headlines-container"
  >
    <div
      class="news-headline"
    >
      <a
        href="https://ca.sports.yahoo.com/news/terence-crawford-shows-hes-level-one-sided-beatdown-jeff-horn-051253278.html"
        rel="noopener noreferrer"
        target="_blank"
      >
        <div
          class="news-headline-title"
          data-testid="NewsHeadlineComponent_title"
        >
          Terence Crawford shows he's on a level all on his own with one-sided beatdown of Jeff Horn
        </div>
      </a>
      <div
        class="news-headline-date-and-source-container"
      >
        <div
          class="news-headline-date"
        >
          10/06/2018
        </div>
        <div
          class="news-headline-source"
        >
          Yahoo.com
        </div>
      </div>
    </div>
    <div
      class="news-headline"
    >
      <a
        href="https://www.washingtonpost.com/world/european-leaders-are-outraged-by-trump-over-g-7-statement-but-theyre-not-surprised/2018/06/10/e7f54887-9022-4adb-9fb1-092c33cb3e91_story.html"
        rel="noopener noreferrer"
        target="_blank"
      >
        <div
          class="news-headline-title"
          data-testid="NewsHeadlineComponent_title"
        >
          European leaders are outraged by Trump over G-7 statement. But they're not surprised.
        </div>
      </a>
      <div
        class="news-headline-date-and-source-container"
      >
        <div
          class="news-headline-date"
        >
          10/06/2018
        </div>
        <div
          class="news-headline-source"
        >
          The Washington Post
        </div>
      </div>
    </div>
    <div
      class="news-headline"
    >
      <a
        href="https://www.cnn.com/2018/06/10/politics/putin-trump-us-russia-arms-intl/index.html"
        rel="noopener noreferrer"
        target="_blank"
      >
        <div
          class="news-headline-title"
          data-testid="NewsHeadlineComponent_title"
        >
          Putin ready to meet Trump over arms race concerns
        </div>
      </a>
      <div
        class="news-headline-date-and-source-container"
      >
        <div
          class="news-headline-date"
        >
          10/06/2018
        </div>
        <div
          class="news-headline-source"
        >
          CNN
        </div>
      </div>
    </div>
    <div
      class="news-headline"
    >
      <a
        href="https://www.hindustantimes.com/india-news/sco-summit-live-pm-modi-meets-xi-jinping-at-welcome-ceremony-in-china-s-qingdao/story-Xb9qGSR9ol0kXvHJxeGmzL.html"
        rel="noopener noreferrer"
        target="_blank"
      >
        <div
          class="news-headline-title"
          data-testid="NewsHeadlineComponent_title"
        >
          SCO summit LIVE: PM Modi says connectivity with neighbouring nations top priority
        </div>
      </a>
      <div
        class="news-headline-date-and-source-container"
      >
        <div
          class="news-headline-date"
        >
          10/06/2018
        </div>
        <div
          class="news-headline-source"
        >
          Hindustantimes.com
        </div>
      </div>
    </div>
    <div
      class="news-headline"
    >
      <a
        href="http://www.foxnews.com/us/2018/06/10/helicopter-plunges-into-wisconsin-river-killing-man.html"
        rel="noopener noreferrer"
        target="_blank"
      >
        <div
          class="news-headline-title"
          data-testid="NewsHeadlineComponent_title"
        >
          Helicopter plunges into Wisconsin river, killing man
        </div>
      </a>
      <div
        class="news-headline-date-and-source-container"
      >
        <div
          class="news-headline-date"
        >
          10/06/2018
        </div>
        <div
          class="news-headline-source"
        >
          Fox News
        </div>
      </div>
    </div>
    <div
      class="news-headline"
    >
      <a
        href="http://www.foxnews.com/us/2018/06/10/soldier-killed-in-somalia-idd-as-8-year-army-veteran-from-arizona.html"
        rel="noopener noreferrer"
        target="_blank"
      >
        <div
          class="news-headline-title"
          data-testid="NewsHeadlineComponent_title"
        >
          Soldier killed in Somalia ID'd as 8-year Army veteran from Arizona
        </div>
      </a>
      <div
        class="news-headline-date-and-source-container"
      >
        <div
          class="news-headline-date"
        >
          10/06/2018
        </div>
        <div
          class="news-headline-source"
        >
          Fox News
        </div>
      </div>
    </div>
    <div
      class="news-headline"
    >
      <a
        href="http://www.latimes.com/sports/more/la-sp-belmont-cherwa-20180609-story.html"
        rel="noopener noreferrer"
        target="_blank"
      >
        <div
          class="news-headline-title"
          data-testid="NewsHeadlineComponent_title"
        >
          The shadow of American Pharoah lingers over Justify's Triple Crown
        </div>
      </a>
      <div
        class="news-headline-date-and-source-container"
      >
        <div
          class="news-headline-date"
        >
          10/06/2018
        </div>
        <div
          class="news-headline-source"
        >
          Latimes.com
        </div>
      </div>
    </div>
    <div
      class="news-headline"
    >
      <a
        href="https://www.bbc.com/news/uk-politics-44428115"
        rel="noopener noreferrer"
        target="_blank"
      >
        <div
          class="news-headline-title"
          data-testid="NewsHeadlineComponent_title"
        >
          Leave.EU 'downplayed Russian links'
        </div>
      </a>
      <div
        class="news-headline-date-and-source-container"
      >
        <div
          class="news-headline-date"
        >
          10/06/2018
        </div>
        <div
          class="news-headline-source"
        >
          Bbc.com
        </div>
      </div>
    </div>
    <div
      class="news-headline"
    >
      <a
        href="https://mmajunkie.com/2018/06/twitter-reacts-cm-punk-ugly-loss-mike-jackson-ufc-225"
        rel="noopener noreferrer"
        target="_blank"
      >
        <div
          class="news-headline-title"
          data-testid="NewsHeadlineComponent_title"
        >
          Twitter reacts to CM Punk's ugly, bloody loss to Mike Jackson at UFC 225
        </div>
      </a>
      <div
        class="news-headline-date-and-source-container"
      >
        <div
          class="news-headline-date"
        >
          10/06/2018
        </div>
        <div
          class="news-headline-source"
        >
          Mmajunkie.com
        </div>
      </div>
    </div>
    <div
      class="news-headline"
    >
      <a
        href="https://www.cnn.com/2018/06/09/us/ice-undocumented-immigrant-mayor-cuomo-letter/index.html"
        rel="noopener noreferrer"
        target="_blank"
      >
        <div
          class="news-headline-title"
          data-testid="NewsHeadlineComponent_title"
        >
          NYC pizza delivery man given emergency stay after immigration detention
        </div>
      </a>
      <div
        class="news-headline-date-and-source-container"
      >
        <div
          class="news-headline-date"
        >
          10/06/2018
        </div>
        <div
          class="news-headline-source"
        >
          CNN
        </div>
      </div>
    </div>
    <div
      class="news-headline"
    >
      <a
        href="https://www.gamespot.com/articles/e3-2018-ea-play-press-conference-shows-the-loot-bo/1100-6459554/"
        rel="noopener noreferrer"
        target="_blank"
      >
        <div
          class="news-headline-title"
          data-testid="NewsHeadlineComponent_title"
        >
          E3 2018: EA Play Press Conference Shows The Loot Box Backlash Paid Off
        </div>
      </a>
      <div
        class="news-headline-date-and-source-container"
      >
        <div
          class="news-headline-date"
        >
          10/06/2018
        </div>
        <div
          class="news-headline-source"
        >
          Gamespot.com
        </div>
      </div>
    </div>
    <div
      class="news-headline"
    >
      <a
        href="https://pagesix.com/2018/06/09/anthony-bourdain-was-regularly-suicidal-after-end-of-first-marriage/"
        rel="noopener noreferrer"
        target="_blank"
      >
        <div
          class="news-headline-title"
          data-testid="NewsHeadlineComponent_title"
        >
          Anthony Bourdain was 'regularly suicidal' after end of first marriage
        </div>
      </a>
      <div
        class="news-headline-date-and-source-container"
      >
        <div
          class="news-headline-date"
        >
          09/06/2018
        </div>
        <div
          class="news-headline-source"
        >
          Pagesix.com
        </div>
      </div>
    </div>
    <div
      class="news-headline"
    >
      <a
        href="http://www.businessinsider.com/chinese-hackers-reportedly-stole-a-bunch-of-us-undersea-warfare-data-2018-6"
        rel="noopener noreferrer"
        target="_blank"
      >
        <div
          class="news-headline-title"
          data-testid="NewsHeadlineComponent_title"
        >
          Chinese hackers reportedly stole a bunch of highly sensitive undersea warfare data from a US Navy contractor
        </div>
      </a>
      <div
        class="news-headline-date-and-source-container"
      >
        <div
          class="news-headline-date"
        >
          09/06/2018
        </div>
        <div
          class="news-headline-source"
        >
          Business Insider
        </div>
      </div>
    </div>
    <div
      class="news-headline"
    >
      <a
        href="https://www.reuters.com/article/us-g7-summit/trump-torpedoes-g7-effort-to-ease-trade-spat-threatens-auto-tariffs-idUSKCN1J50H3"
        rel="noopener noreferrer"
        target="_blank"
      >
        <div
          class="news-headline-title"
          data-testid="NewsHeadlineComponent_title"
        >
          Trump torpedoes G7 effort to ease trade spat, threatens auto tariffs
        </div>
      </a>
      <div
        class="news-headline-date-and-source-container"
      >
        <div
          class="news-headline-date"
        >
          09/06/2018
        </div>
        <div
          class="news-headline-source"
        >
          Reuters
        </div>
      </div>
    </div>
    <div
      class="news-headline"
    >
      <a
        href="https://www.cbsnews.com/news/search-teams-find-body-of-woman-missing-in-suspected-alligator-attack-davie-florida/"
        rel="noopener noreferrer"
        target="_blank"
      >
        <div
          class="news-headline-title"
          data-testid="NewsHeadlineComponent_title"
        >
          Search teams find body of woman missing in suspected alligator attack
        </div>
      </a>
      <div
        class="news-headline-date-and-source-container"
      >
        <div
          class="news-headline-date"
        >
          09/06/2018
        </div>
        <div
          class="news-headline-source"
        >
          CBS News
        </div>
      </div>
    </div>
    <div
      class="news-headline"
    >
      <a
        href="http://www.ibtimes.com/mystery-biblical-kings-3000-year-old-sculpted-head-unearthed-2689112"
        rel="noopener noreferrer"
        target="_blank"
      >
        <div
          class="news-headline-title"
          data-testid="NewsHeadlineComponent_title"
        >
          Mystery Biblical King's 3000-Year-Old Sculpted Head Unearthed
        </div>
      </a>
      <div
        class="news-headline-date-and-source-container"
      >
        <div
          class="news-headline-date"
        >
          09/06/2018
        </div>
        <div
          class="news-headline-source"
        >
          Ibtimes.com
        </div>
      </div>
    </div>
    <div
      class="news-headline"
    >
      <a
        href="https://www.washingtonpost.com/world/national-security/in-charging-senate-staffer-and-seizing-reporters-records-justice-dept-ignites-debate-over-leak-crackdown/2018/06/08/6967e630-6b36-11e8-bf8c-f9ed2e672adf_story.html"
        rel="noopener noreferrer"
        target="_blank"
      >
        <div
          class="news-headline-title"
          data-testid="NewsHeadlineComponent_title"
        >
          In charging Senate staffer and seizing reporter's records, Justice Dept. ignites debate over leak crackdown
        </div>
      </a>
      <div
        class="news-headline-date-and-source-container"
      >
        <div
          class="news-headline-date"
        >
          08/06/2018
        </div>
        <div
          class="news-headline-source"
        >
          The Washington Post
        </div>
      </div>
    </div>
    <div
      class="news-headline"
    >
      <a
        href="https://www.washingtonpost.com/powerpost/youve-handed-us-an-issue-democrats-pounce-on-trump-administrations-health-care-move/2018/06/08/643505f0-6b34-11e8-bea7-c8eb28bc52b1_story.html"
        rel="noopener noreferrer"
        target="_blank"
      >
        <div
          class="news-headline-title"
          data-testid="NewsHeadlineComponent_title"
        >
          'You've handed us an issue': Democrats pounce on Trump administration's health-care move
        </div>
      </a>
      <div
        class="news-headline-date-and-source-container"
      >
        <div
          class="news-headline-date"
        >
          08/06/2018
        </div>
        <div
          class="news-headline-source"
        >
          The Washington Post
        </div>
      </div>
    </div>
  </div>
  <div
    class="news-show-more-button"
  >
    Show More
  </div>
  <div
    class="news-powered-by"
  >
    <a
      href="https://newsapi.org"
      rel="noopener noreferrer"
      target="_blank"
    >
      Powered by News API
    </a>
  </div>
</div>
`;

exports[`NewsHeadlineComponent renders correctly 1`] = `
<div
  class="news-headline"
>
  <a
    href="https://www.cnn.com/2018/06/10/politics/putin-trump-us-russia-arms-intl/index.html"
    rel="noopener noreferrer"
    target="_blank"
  >
    <div
      class="news-headline-title"
      data-testid="NewsHeadlineComponent_title"
    >
      Putin ready to meet Trump over arms race concerns
    </div>
  </a>
  <div
    class="news-headline-date-and-source-container"
  >
    <div
      class="news-headline-date"
    >
      10/06/2018
    </div>
    <div
      class="news-headline-source"
    >
      CNN
    </div>
  </div>
</div>
`;
